File: Assets/Scripts/MazeGen/Cluster.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using System;

public enum ClusterShape
{
    Plus,
    TeeNorth,
    TeeSouth,
    TeeEast,
    TeeWest,
    CornerNorthEast,
    CornerSouthEast,
    CornerNorthWest,
    CornerSouthWest,
    VerticalLine,
    HorizontalLine,
    Empty
}
public class Cluster
{
    public Coordinate Location { get; set; }
    public List<ClusterShape> PossibleShapes { get; private set; }
    public ClusterShape DefiniteShape { get; private set; }

    Dictionary<ClusterDirection, Cluster> neighbors;

    
    public bool IsDefiniteShapeSet { get; private set; }

    public int Entropy { get { return PossibleShapes.Count; } }   

    // Start is called before the first frame update
    public Cluster()
    {
        // initialize by adding all possible shapes
        PossibleShapes = new List<ClusterShape>((ClusterShape[])Enum.GetValues(typeof(ClusterShape)));
        IsDefiniteShapeSet = false;
        neighbors = new Dictionary<ClusterDirection, Cluster>();
    }

    public void SetNeighbor(Cluster cluster, ClusterDirection direction)
    {
        neighbors[direction] = cluster;
    }

    public Cluster Neighbor(ClusterDirection direction)
    {
        return neighbors[direction];
    }

    public void ConstrainDirection(ClusterShape[] shapes)
    {
        if (PossibleShapes.Count == 1)
            return;
        // if the shapes are in this group's list of available, or possible, shapes, remove it/them
        foreach (ClusterShape shape in shapes)
        {
            if (PossibleShapes.Contains(shape))
            {
                PossibleShapes.Remove(shape);
            }
        }
    }

    public void SetDefiniteShape(ClusterShape shape)
    {
        DefiniteShape = shape;

        if (PossibleShapes == null)
            PossibleShapes = new List<ClusterShape>();

        PossibleShapes.Clear();
        PossibleShapes.Add(shape);
        IsDefiniteShapeSet = true;
        DefiniteShape = shape;
    }

    public void SetPossibleShapes(List<ClusterShape> shapes)
    {
        if (PossibleShapes == null)
        {
            PossibleShapes = new List<ClusterShape>();
        }
        PossibleShapes.Clear();
        PossibleShapes.AddRange(shapes);
    }

    public bool FilterToCompatibleShapes(ClusterShape[] shapes, ClusterDirection dir)
    {
        // for each of our existing shapes, check if ANY of the incoming shapes are compatible.
        // otherwise remove 
        bool isReduced = false;

        foreach (ClusterShape myShape in shapes)
        {
            bool isCompatible = false;
            // is this shape compatible with any of the incoming shapes?
            // 
            
        }

        if (shapes != null)
        {
            foreach (ClusterShape shape in shapes)
            {

            }
        }
        return isReduced;
    }


    public bool ReducePossibleShapes(ClusterShape[] shapes)
    {
        bool isReduced = false;

        if (shapes != null)
        {
            foreach (ClusterShape shape in shapes)
            {
                if (PossibleShapes.Contains(shape))
                {
                    PossibleShapes.Remove(shape);
                    isReduced = true;
                }
            }
        }

        if (Entropy == 1)
        {
            SetDefiniteShape(PossibleShapes[0]);
        }

        return isReduced;
    }

    public void SetRandomDefiniteShape()
    {
        ClusterShape shape = PossibleShapes[UnityEngine.Random.Range(0, PossibleShapes.Count)];

        SetDefiniteShape(shape);
        PossibleShapes.Clear();
        PossibleShapes.Add(shape);
    }

    // returns clusers to north, east, south, west, in that order.
    // null if no neighbor
}

File: lua/cluster.lua
ClusterShape = {
    Plus = "Plus",
    TeeNorth = "TeeNorth",
    TeeSouth = "TeeSouth",
    TeeEast = "TeeEast",
    TeeWest = "TeeWest",
    CornerNorthEast = "CornerNorthEast",
    CornerSouthEast = "CornerSouthEast",
    CornerNorthWest = "CornerNorthWest",
    CornerSouthWest = "CornerSouthWest",
    VerticalLine = "VerticalLine",
    HorizontalLine = "HorizontalLine",
    Empty = "Empty"
}

Cluster = {}
Cluster.__index = Cluster

function Cluster:new()
    local instance = setmetatable({}, Cluster)
    instance.Location = nil
    instance.PossibleShapes = {}
    for _, shape in pairs(ClusterShape) do
        table.insert(instance.PossibleShapes, shape)
    end
    instance.DefiniteShape = nil
    instance.neighbors = {}
    instance.IsDefiniteShapeSet = false
    return instance
end

function Cluster:SetNeighbor(cluster, direction)
    self.neighbors[direction] = cluster
end

function Cluster:Neighbor(direction)
    return self.neighbors[direction]
end

function Cluster:ConstrainDirection(shapes)
    if #self.PossibleShapes == 1 then
        return
    end
    for _, shape in ipairs(shapes) do
        for i = #self.PossibleShapes, 1, -1 do
            if self.PossibleShapes[i] == shape then
                table.remove(self.PossibleShapes, i)
            end
        end
    end
end

function Cluster:SetDefiniteShape(shape)
    self.DefiniteShape = shape
    self.PossibleShapes = {shape}
    self.IsDefiniteShapeSet = true
end

function Cluster:SetPossibleShapes(shapes)
    self.PossibleShapes = {}
    for _, shape in ipairs(shapes) do
        table.insert(self.PossibleShapes, shape)
    end
end

function Cluster:FilterToCompatibleShapes(shapes, dir)
    local isReduced = false
    -- Logic for filtering compatible shapes would go here
    return isReduced
end

function Cluster:ReducePossibleShapes(shapes)
    local isReduced = false
    if shapes then
        for _, shape in ipairs(shapes) do
            for i = #self.PossibleShapes, 1, -1 do
                if self.PossibleShapes[i] == shape then
                    table.remove(self.PossibleShapes, i)
                    isReduced = true
                end
            end
        end
    end
    if #self.PossibleShapes == 1 then
        self:SetDefiniteShape(self.PossibleShapes[1])
    end
    return isReduced
end

function Cluster:SetRandomDefiniteShape()
    local index = math.random(1, #self.PossibleShapes)
    local shape = self.PossibleShapes[index]
    self:SetDefiniteShape(shape)
end

-- returns clusters to north, east, south, west, in that order.
-- nil if no neighbor
